automated-testing/ui-testing/readme.md

```diff
@@ -0,0 +1,64 @@
+
+## Evidence and Measures
+
+Integrating UI Tests in to your CI/CD is necessary but more challenging than unit tests.  The increased challenge is that UI tests either need to run headlessly with something like [Puppeteer](https://github.com/puppeteer/puppeteer) or there needs to be more extensive orchestration with Azure DevOps or Github that would handle the full testing integration for you like [BrowserStack](https://www.browserstack.com/automate/azure)
+
+Integrations like `BrowserStack` are nice since they provide Azure DevOps reports as part of the test run.
+
+That said, Azure DevOps supports a variety of test adapters, so you can use any UI Testing framework that supports outputting the test results to one of the output formats listed at [Publish Test Results task](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/test/publish-test-results?view=azure-devops&tabs=yaml)
+
+## General Guidance
+
+The scope of UI testing should be strategic. UI tests can take a significant amount of time to both implement and run, and it challenging to test every possibility of user interaction in a production application due to the large possible number of possible interactions.
+
+Designing the UI tests around the functional tests makes sense.  For example, given an input form, a UI test would ensure that the visual representation is consistent across devices, is accessible and interactable, and is consistent across code changes.
+
+UI Tests will catch 'runtime' bugs that unit and functional tests won't.  For example if the submit button for an input form is rendered but not clickable due to a positioning bug in the UI, then this could be considered a runtime bug that would not have been caught by unit or functional tests.
+
+UI Tests can run on mock data or snapshots of production data, like in QA or staging.
+
+### Writing Tests
+
+Good UI tests follow a few general principles:
+
+- Choose a UI testing framework that enables quick feedback and is easy to use
+- Design the UI to be easily testable.  For example, add CSS selectors or set the id on elements in a web page to allow easier selecting.
+- Test on all primary devices that the user uses, don't just test on a single device or OS.
+- When a test mutates data ensure that data is created on demand and cleaned up after.  The consequence of not doing this would be inconsistent testing.  
+
+### Gotchas
+
+UI Testing can get very challenging at the lower level, especially with a testing framework like Selenium.  If you choose to go this route, then you'll likely encounter timeouts, missing elements, and you'll have significant friction with the testing framework itself.  Due to many issues with UI testing there have been a number of free and paid solutions that help alleviate certain issues with frameworks like Selenium.  This is why you'll find Cypress in the recommended frameworks as it solves many of the known issues with Selenium.
```

Comment:
* I would advise against criticizing one tool vs the other, we can talk about pitfalls in general and maybe in another section, we can mention some commonly used tools and their pros/cons 
* Can we link Selenium to the tool website https://www.selenium.dev/

Reply:
While I agree with you in general, Selenium is a huge challenge to work with by itself.  Entire ecosystems and frameworks have been built up because of how challenging it is to work with.  I'd much rather make that very apparent to would be first time selenium users than something more gentle.

That said, I think it makes sense not to endorse a specific framework here like Cypress, but to just point out that there are many solutions that either build on top of Selenium or that are designed from the ground up to solve issues that Selenium has.

Will update the wording here and not resolve, feel free to continue the conversation and I can iterate the wording as necessary.